implementations/rust/ockam/ockam_command/src/util/github.rs

```diff
@@ -0,0 +1,130 @@
+use std::{fs::File, io::Write, path::PathBuf};
+
+use clap::crate_version;
+use colorful::Colorful;
+use miette::{miette, IntoDiagnostic};
+use serde::Deserialize;
+use tokio::runtime::Builder;
+
+use crate::error::Result;
+
+#[derive(Deserialize, Debug)]
+pub struct LatestRelease {
+    name: String,
+}
+
+impl LatestRelease {
+    pub fn version(&self) -> Result<&str> {
+        let result = self.name.split_once('v');
+        match result {
+            Some((_, version)) => Ok(version),
+            None => Err(miette!("Failed to get latest release version").into()),
+        }
+    }
+}
+
+pub fn get_latest_release_version_sync() -> Result<LatestRelease> {
+    Builder::new_current_thread()
+        .enable_all()
+        .build()
+        .unwrap()
+        .block_on(get_latest_release_version())
+}
+
+pub async fn get_latest_release_version() -> Result<LatestRelease> {
+    let url = "https://api.github.com/repos/build-trust/ockam/releases/latest";
+    let client = reqwest::Client::new();
+    let resp = client
+        .get(url)
+        .header("Accept", "application/vnd.github+json")
+        .header("X-GitHub-Api-Version", "2022-11-28")
+        .header("User-Agent", "Ockam Command")
+        .send()
+        .await;
+    if let Ok(r) = resp {
+        if let Ok(release) = r.json::<LatestRelease>().await {
+            return Ok(release);
+        }
+    }
+    Err(miette!("Failed to get latest release").into())
+}
+
+pub fn download_install_file_sync(install_path: &PathBuf) -> miette::Result<()> {
+    Builder::new_current_thread()
+        .enable_all()
+        .build()
+        .unwrap()
+        .block_on(download_install_file(install_path))
+}
+
+pub async fn download_install_file(install_path: &PathBuf) -> miette::Result<()> {
+    let install_script_url =
+        "https://raw.githubusercontent.com/build-trust/ockam/develop/install.sh";
+
+    let client = reqwest::Client::new();
+    let response = client
+        .get(install_script_url)
+        .send()
+        .await
+        .into_diagnostic()?;
+    let mut file = File::create(install_path).into_diagnostic()?;
+    file.write_all(&response.bytes().await.into_diagnostic()?)
+        .into_diagnostic()
+}
+
+#[derive(Deserialize, Debug)]
+struct UpgradeFile {
+    upgrade_message: Option<String>,
+    upgrade_message_macos: Option<String>,
+}
+
+pub fn check_upgrade_sync() {
+    Builder::new_current_thread()
+        .enable_all()
+        .build()
+        .unwrap()
+        .block_on(check_upgrade());
+}
+
+pub async fn check_upgrade() {
+    let url = format!(
+        "https://github.com/build-trust/ockam/releases/download/ockam_v{}/upgrade.json",
+        crate_version!()
+    );
+    let resp = reqwest::get(url).await;
+
+    if let Ok(r) = resp {
+        if let Ok(upgrade) = r.json::<UpgradeFile>().await {
+            if let Some(message) = upgrade.upgrade_message {
+                eprintln!("\n{}", message.yellow());
```

Comment:
We need to use the `opts.terminal` instead of `eprintln` here.

Reply:
This was also part of the original code that checked for the latest version from the `version.json` file on GitHub, I just moved it to this file. I can refactor this as part of this PR as well if you would like.

Reply:
> I can refactor this as part of this PR as well if you would like.

I see. Yes that would be nice. We are trying to move away from bare `printlns` in order to completely control the rendering in the terminal.

Reply:
> I could also rework this to not use the version.json file and instead just pull the latest version from the GitHub releases?

I don't understand that part. Are you talking about the `upgrade.json` file? I actually tried to download it from `https://github.com/build-trust/ockam/releases/download/ockam_v0.90.0/upgrade.json` and it comes back empty. Is that the same for you too? Because then this means that we also need to fix that part somewhere in our release process (FYI @metaclips)

Reply:
Yeah exactly. I mentioned it in the description for this PR, but I found that the `upgrade.json` were never really up-to-date. You can play around with it by changing the `version` in the `Cargo.toml` file. Currently I only no suggestions for `0.89.0`, `0.88.0` and `0.87.0`. If I set it to `0.86.0` then it informs me that there is an update available, but that the latest version is `0.87.0`.
<img width="869" alt="image" src="https://github.com/build-trust/ockam/assets/22850972/3687e948-6aee-4f4e-88ec-91d48f88411b">
Currently it pulls this information from the release of the version that the user is using, meaning these files would need to be updated every release to be effective. 
That's why I ended up just pulling the latest version from the release instead of just using the `version.json` file.